WRF: Add version 3.9.1.1 and improve recipe robustness

ptooley opened the pull request:
PR to improve several aspects of WRF support in Spack:

The headline issue here is the fact that upstream use `make -i -k` and their custom make script always returns 0. This means that the standard approach of checking the return code won't detect a failed build.  Instead we need to actually check the output of the script to see if it succeeded.  My approach is the least-bad solution I could find.

Second big issue is that the configure script *enforces* interactivity and dynamically generates its option.  Previous recipe version had response values hardcoded for an assumed x86 platform.  This new version should be platform independent and has been tested on at least aarch64 and x86_64.

Additionally I have added support for v 3.9 which is a common HPC benchmarking workload, and fixed an issue with failure to compile against recent glibcs that don't ship sunrpc headers.

* Include version 3.9.1.1 as common benchmarking workload
* Fix compilation against recent glibc (use spack installed libtirpc)
* Detect and handle failed compilation (upstream use make -i)

ptooley committed:
Add WRF 3.9.1.1 and improve recipe robustness

* Include version 3.9.1.1 as common benchmarking workload
* Fix compilation against recent glibc (detect spack installed libtirpc)
* Detect and handle failed compilation (upstream use make -i)

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,309 @@ class Wrf(Package):
     for both atmospheric research and operational forecasting applications.
     """
 
-    homepage = "https://www.mmm.ucar.edu/weather-research-and-forecasting-model"
-    url      = "https://github.com/wrf-model/WRF/archive/v4.2.tar.gz"
-    maintainers = ['MichaelLaufer']
+    homepage = (
+        "https://www.mmm.ucar.edu/weather-research-and-forecasting-model"
+    )
+    url = "https://github.com/wrf-model/WRF/archive/v4.2.tar.gz"
+    maintainers = ["ptooley"]
```
Can you add your name to the list of maintainers instead of replacing @MichaelLaufer?

ptooley replied:
Yes, sorry. This started life as a new recipe because the changes were so big.

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,309 @@ class Wrf(Package):
+    patch("patches/4.2/tirpc_detect.patch", when="@4.2")
+    patch("patches/4.2/add_aarch64.patch", when="@4.2")
+
+    depends_on("pkg-config", type=("build"))
```
```suggestion
    depends_on("pkgconfig", type=("build"))
```
Spack has a virtual package that lets you use `pkg-config` or `pkgconf`

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,309 @@ class Wrf(Package):
+                csh_bin,
+                "./compile",
+                "-j",
+                "10",
```
I assume this is the level of parallelism? In that case:
```suggestion
                str(make_jobs),
```

ptooley replied:
This is result of experiments where I was still getting seemingly random compilation failures with more than this many build jobs.  The previous recipe was using up to 20 but that was still not working well for me.  I'll reinstate the up-to-twenty logic for now...

adamjstewart replied:
You can also use up-to-10 if that helps. Did I also see you use `parallel=False`? I think that would cap at 1.

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,309 @@ class Wrf(Package):
+    def run_compile_script(self):
+        csh_bin = path_join(self.spec["tcsh"].prefix, "bin/csh")
+
+        compilation = Popen(
```
Instead of using `Popen`, is it possible to use Spack's `Executable` class? Something like:
```python
csh = Executable(csh_bin)
csh(...)
```

ptooley replied:
Having looked at the source for the Executable class I think I can, but the text output from that compile stage is massive, so I was trying to avoid keeping the whole build log in RAM.  If spack is doing that anyway behind the scenes then I think I can convert this one to `Executable` but not the interactive configure step.

adamjstewart replied:
Ah, I don't think we're doing that, didn't realize this package needed to be so complicated.

ptooley replied:
Testing with `Executable` gives me the behaviour I need, thanks!

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,309 @@ class Wrf(Package):
+            raise InstallError("Configure failed - unknown error")
+
+    def run_compile_script(self):
+        csh_bin = path_join(self.spec["tcsh"].prefix, "bin/csh")
```
```suggestion
        csh_bin = self.spec["tcsh"].prefix.bin.csh
```

adamjstewart reviewed (changes requested):
@MichaelLaufer can you review this?

ptooley committed:
WRF: PR changes round 1
fix build jobs
fix maintainers
fix pkgconfig dependency
use Executable to run compile stage
repair some overzealous autoformatting by black

ptooley commented:
All changes included and I also cleaned up some of the more egregious formatting choices my autoformatter made...

Apologies for the force push - I tried to rebase my local repo onto the latest develop and got in a complete mess.

ptooley committed:
WRF: make recipe py26 compatible

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,269 @@ class Wrf(Package):
+        if "Please select from among the following" in outputbuf:
+            options = collect_platform_options(outputbuf)
+            comp_pair = "%s/%s" % (
+                basename(self.compiler.fc),
+                basename(self.compiler.cc),
```
`self.compiler.cc` and friends are the actual compilers, generally we want to use Spack's compiler wrappers (`spack_cc`) so that we automatically inject the appropriate linking args. Does it not work if you use Spack's compiler wrappers?

ptooley replied:
Here I am picking up the actual name of the compiler binary to compare with the options presented by the interactive configure script.  It doesn't actually inject any paths.

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,269 @@ class Wrf(Package):
+
+    def do_configure_fixup(self):
+        # Fix mpi compiler wrapper aliases
+        if (
+            self.spec.satisfies("@3.9.1.1")
+            and self.spec.compiler.name == "gcc"
+        ):
```
```suggestion
        if self.spec.satisfies("@3.9.1.1 %gcc"):
```

ptooley replied:
Done

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,269 @@ class Wrf(Package):
+            ) as ofh:
+                for line in ifh:
+                    if line.startswith("DM_"):
+                        line = line.replace("mpif90 -f90=$(SFC)", "mpif90")
+                        line = line.replace("mpicc -cc=$(SCC)", "mpicc")
```
We could use `self.spec['mpi'].mpicc` and friends here instead, that may be more robust as not every MPI library uses the same names.

ptooley replied:
Done

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,269 @@ class Wrf(Package):
         # num of compile jobs capped at 20 in wrf
-        csh('./compile', '-j', str(min(int(make_jobs), 20)),
-            spec.variants['compile_type'].value)
+        num_jobs = str(make_jobs if make_jobs < 20 else 20)
```
I think the old logic of `str(min(int(make_jobs), 20))` is more clear.

ptooley replied:
Done

adamjstewart reviewed (commented):
Few more comments on the Spack recipe. I don't know much about WRF so I can't review the patches.

ptooley committed:
wrf: recipe review changes round 2

adamjstewart reviewed (approved):
Looks good to me from my end. Since this is a pretty big change I'm still hoping for a review from @MichaelLaufer or another WRF user. Let's give it a few days. If we don't hear back, ping me and I'll merge.

adamjstewart commented on var/spack/repos/builtin/packages/wrf/package.py:
```diff
@@ -15,132 +60,270 @@ class Wrf(Package):
+                "./arch/configure_new.defaults",
+                "./arch/configure_new.defaults.bak",
+            )
+            with open("./arch/configure_new.defaults.bak", "rt") as ifh, open(
+                "./arch/configure_new.defaults", "wt"
+            ) as ofh:
```
The unit tests are failing on this line. I don't think you can open two file handles on the same line. You should doubly nest this.

ptooley replied:
Yeah Python 2.6 chokes on this, 2.7 onwards it is okay. Vermin didn't flag it and I can't even install Python2.6 from my package manager to test.

ptooley committed:
more python 26 fixes

ptooley commented:
> Looks good to me from my end. Since this is a pretty big change I'm still hoping for a review from @MichaelLaufer or another WRF user. Let's give it a few days. If we don't hear back, ping me and I'll merge.

Thanks! Do you want this squashed to a single commit before merge?

adamjstewart commented:
> Do you want this squashed to a single commit before merge?

Nah, we automatically squash and merge all PRs, so there's no need to do it yourself.

MichaelLaufer commented:
> 
> 
> Looks good to me from my end. Since this is a pretty big change I'm still hoping for a review from @MichaelLaufer or another WRF user. Let's give it a few days. If we don't hear back, ping me and I'll merge.

@ptooley @adamjstewart 
We are having some IT issues on our cluster at the moment so I could not test this out. I hope to have it fixed by tomorrow. Looks good, and I am happy to see V3.9.1.1 support!

adamjstewart commented:
In that case I'll merge for now and we can backport any changes if they are needed later. Thanks for reviewing @MichaelLaufer!
